### PR TITLE
Common Bft forks schedule

### PR DESCRIPTION
**Pull request by jframe**
## PR description
Create a common bft forks schedule that can be used between both ibft and qbft. This is necessary to add a bft transition for block period.

## Changelog

- [x] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: config/src/main/java/org/hyperledger/besu/config/JsonBftConfigOptions.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+  @Override
+  public BigInteger getBlockRewardWei() {
+    final Optional<String> configFileContent = JsonUtil.getString(bftConfigRoot, "blockreward");
+
+    if (configFileContent.isEmpty()) {
+      return BigInteger.ZERO;
+    }
+    final String weiStr = configFileContent.get();
+    if (weiStr.startsWith("0x")) {
```
[the remark this answers is not in the document]
Comment (jframe) [reply config/src/main/java/org/hyperledger/besu/config/JsonBftConfigOptions.java:101]: That shouldn't be necessary, all config keys are normalised as lowercase https://github.com/hyperledger/besu/blob/main/config/src/main/java/org/hyperledger/besu/config/GenesisConfigFile.java#L85

[the remark this answers is not in the document]
Comment (jframe) [reply config/src/main/java/org/hyperledger/besu/config/JsonBftConfigOptions.java:101]: Gotcha, yeah I think it should

[the remark this answers is not in the document]
Comment (jframe) [reply config/src/main/java/org/hyperledger/besu/config/JsonBftConfigOptions.java:101]: done

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validator/ForkingValidatorProvider.java [hunk 2]
```diff
@@ -71,46 +70,32 @@ public ForkingValidatorProvider(
...
-        if (block > 0 && block == fork.get().getBlock()) {
-          final long prevBlockNumber = block - 1L;
-          final Optional<BlockHeader> prevBlockHeader = blockchain.getBlockHeader(prevBlockNumber);
-          if (prevBlockHeader.isPresent()) {
-            return resolveValidatorProvider(prevBlockNumber)
-                .getValidatorsForBlock(prevBlockHeader.get());
-          }
+    // when moving to a block validator the first block needs to be initialised or created with
+    // the previous block state otherwise we would have no validators
+    if (forkSpec.getConfigOptions().getValidatorContractAddress().isEmpty()) {
```
Comment (siladu) [inline consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validator/ForkingValidatorProvider.java:78]: This check feels slightly less obvious than before.
Not a blocker, just a suggestion:
Add isBlockHeaderMode() (and isValidatorContractMode()?) for clarity?
Or maybe validatorSelectionMode.equals(BLOCKHEADER) as before?

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validator/ForkingValidatorProvider.java [hunk 3]
```diff
@@ -33,13 +32,13 @@
 public class ForkingValidatorProvider implements ValidatorProvider {
 
   private final Blockchain blockchain;
-  private final ValidatorSelectorForksSchedule forksSchedule;
+  private final BftForksSchedule<QbftConfigOptions> forksSchedule;
```
Comment (siladu) [inline consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validator/ForkingValidatorProvider.java:35]: The type parameter now couples Qbft to ForkingValidatorProvider...is that a problem or is it this class likely to remain in the qbft package?

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/MutableQbftConfigOptions.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+ *  SPDX-License-Identifier: Apache-2.0
+ */
+package org.hyperledger.besu.consensus.qbft;
+
+import org.hyperledger.besu.config.QbftConfigOptions;
+import org.hyperledger.besu.consensus.common.MutableBftConfigOptions;
+
+import java.util.Optional;
+
+public class MutableQbftConfigOptions extends MutableBftConfigOptions implements QbftConfigOptions {
```
Comment (siladu) [inline consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/MutableQbftConfigOptions.java:22]: I like the structure and naming of the Mutable/Json ConfigOptions. It wasn't immediately clear to me how/why it was necessary.
I guess the need to mutate the config options is specific to transitions, so I wonder if a short sentence at the top of MutableQbftConfigOptions/JsonQbftConfigOptions would help shortcut a future reader's understanding?

**Review by siladu — APPROVED** [review]
Seems like a good change to me and also cleans up some of the client code. Some questions and suggestions but nothing major.

File: config/src/main/java/org/hyperledger/besu/config/BftConfigOptions.java [hunk 5]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright ConsenSys AG.
+ * Copyright Hyperledger Besu Contributors.
```
[the remark this answers is not in the document]
Comment (jframe) [reply config/src/main/java/org/hyperledger/besu/config/BftConfigOptions.java:2]: done

**Pull request merged**